### PR TITLE
add fan speeds table

### PR DESCRIPTION
**Pull request by kaizensoze**
Add a fan_speeds table with each row containing attributes for a given
fan such as min, max, actual speed.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/tables/system/darwin/smc_keys.cpp [hunk 1]
```diff
@@ -490,5 +497,72 @@ QueryData getTemperatures(QueryContext &context) {
...
+    r["fan"] = std::to_string(fanIdx);
+
+    for (const auto &smcFanSpeedKey : kSMCFanSpeeds) {
+      std::stringstream key;
+      key << boost::format(smcFanSpeedKey.first) % fanIdx;
+
+      QueryData key_data;
+
+      genSMCKey(key.str(), smc, key_data);
+      if (!key_data.empty()) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/smc_keys.cpp]: how about:
```cpp
if (key_data.empty()) {
  continue;
}

auto &smcRow = key_data.back();
if (smcRow["value"].empty()) {
  continue;
}
```

File: osquery/tables/system/darwin/smc_keys.cpp [hunk 2]
```diff
@@ -490,5 +497,72 @@ QueryData getTemperatures(QueryContext &context) {
...
+          }
+
+          int fanSpeed = (int(val[0]) << 6) + (int(val[1]) >> 2);
+          r[smcFanSpeedKey.second] = std::to_string(fanSpeed);
+        }
+      }
+    }
+
+    // Skip adding fan if no useful data.
+    if (r.size() == 1) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/smc_keys.cpp]: If you prefer to `continue` on each error, you wont need this final check.

File: specs/darwin/fan_speeds.table [hunk 3]
```diff
@@ -0,0 +1,11 @@
+table_name("fan_speeds")
+description("Fan speeds.")
+schema([
+    Column("fan", TEXT, "Fan number"),
```
Comment (theopolis) [inline specs/darwin/fan_speeds.table]: If these are all ints you can use `INTEGER` for the type.

File: osquery/tables/system/darwin/smc_keys.cpp [hunk 4]
```diff
@@ -490,5 +497,72 @@ QueryData getTemperatures(QueryContext &context) {
...
+          } catch (const boost::algorithm::hex_decode_error &e) {
+            continue;
+          }
+
+          if (val.size() != 2) {
+            continue;
+          }
+
+          int fanSpeed = (int(val[0]) << 6) + (int(val[1]) >> 2);
+          r[smcFanSpeedKey.second] = std::to_string(fanSpeed);
```
Comment (theopolis) [inline osquery/tables/system/darwin/smc_keys.cpp]: We have a helper macro `INTEGER()` that's preferred over `std::to_string` *only* when assigning column values. The macro will help if/when we convert `Row`s to dynamics and can support the literal types via a union or something smart. ;)

File: specs/darwin/fan_speeds.table (hunk 3 above)
Comment (kaizensoze) [inline specs/darwin/fan_speeds.table]: Should an integer column default to -1 if there's no value?

File: osquery/tables/system/darwin/smc_keys.cpp (hunk 2 above)
Comment (kaizensoze) [inline osquery/tables/system/darwin/smc_keys.cpp]: Even if every iteration of the inner for loop fails, you'll still have the r["fan"] set from the outer for loop, which is useless by itself, so I think you still need the continue in the outer loop, no?

**Comment by guerrillera** [comment]
Muchisimas gracias a todos , tengo que hablar con github con verdadera urgencia

File: specs/darwin/fan_speeds.table (hunk 3 above)
Comment (theopolis) [inline specs/darwin/fan_speeds.table]: It's mostly up to you to define, then capture the "failed"/"unknown" state in the column description within the spec.

File: osquery/tables/system/darwin/smc_keys.cpp (hunk 2 above)
Comment (theopolis) [inline osquery/tables/system/darwin/smc_keys.cpp]: I'd say there's some value in having the the `fan` key with no values. It'll be valuable debug information-- guessing it's a fairly rare state.

Comment (kaizensoze) [inline osquery/tables/system/darwin/smc_keys.cpp]: Yeah good point.

**Comment by kaizensoze** [comment]
Updated. (I always forget if a rebase + force push to a PR branch triggers a notification.)

File: osquery/tables/system/darwin/smc_keys.cpp [hunk 5]
```diff
@@ -490,5 +496,73 @@ QueryData getTemperatures(QueryContext &context) {
...
+  }
+
+  // Get attributes for each fan.
+  int numFans = std::stoi(smcRow["value"]);
+  for (int fanIdx = 0; fanIdx < numFans; fanIdx++) {
+    Row r;
+    r["fan"] = std::to_string(fanIdx);
+
+    for (const auto &smcFanSpeedKey : kSMCFanSpeeds) {
+      std::stringstream key;
```
Comment (theopolis) [inline osquery/tables/system/darwin/smc_keys.cpp]: Maybe move the `r[smcFanSpeedKey.second] = INTEGER(0);` here, then each error case can happily continue without care :dancers:

File: osquery/tables/system/darwin/smc_keys.cpp [hunk 6]
```diff
@@ -451,8 +457,8 @@ void genTemperature(const std::string &key,
     return;
   }
 
-  int intValue = (val[0] * 256 + val[1]) >> 2;
-  float floatValueCelsius = intValue / 64.0;
+  int intValue = (val[0] * 256 + val[1]);
+  float floatValueCelsius = intValue / 256.0;
```
Comment (theopolis) [inline osquery/tables/system/darwin/smc_keys.cpp]: Note: This will eventually need a rebase against master.

**Comment by theopolis** [comment]
I don't think a force update generates a notification, though if we get this building then the bot will take care of spamming with the build status. Holding off on triggering the build for a few minutes.

File: osquery/tables/system/darwin/smc_keys.cpp (hunk 6 above)
Comment (kaizensoze) [inline osquery/tables/system/darwin/smc_keys.cpp]: Yep I figured #1825, #1826 would get merged first and then I'll rebase against master.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/1827/merge (Job: 2358) is good! :+1:

**Comment by theopolis** [comment]
Hey @kaizensoze feel free to rebase and we'll merge.

**Comment by kaizensoze** [comment]
:+1:

**Comment by osqueryer** [comment]
Build origin/pr/1827/merge (Job: 2359) is good! :+1:

**Pull request merged**